Architecture.md

```diff
@@ -71,6 +71,116 @@
+
+The option we chose is to create a custom git server, which handles the authentication and authorization of the clients, and then execute only the command `git-receive-pack` which is used to push to a repository. This way, the clients can only push to their own repositories, and not to the whole host machine.
+
+The advantage of this solution is that it is lightweight and give us full control over the authentication and authorization of the clients. It also allows us to use the same authentication and authorization system as the rest of the application, which is easier to maintain.
+
+The git server is written in golang, for its simplicity and its performance. It is also well documented, and has a lot of libraries to help us build the server, unlike Rust which was the original choice. The libraries used are [ssh made by gliderlabs](https://github.com/gliderlabs/ssh) which is used to create git clients and servers. It is also used by [gitea](https://about.gitea.com) to create their git server.
+
+The authentication and authorization process is the following:
+
+- The client connects to the git server using ssh via `git push`
+- The git server checks if the client is authenticated using the ssh key provided by the client by matching it against the ssh keys stored in the database
+    - If the client is not authenticated, the connection is closed
+- The git server checks if the client is authorized to push to the repository by checking if the client is the owner of the repository
+    - If the client is not authorized, the connection is closed
+- The git server executes the command `git-receive-pack` which is used to push to a repository
+
+```mermaid
+sequenceDiagram
+    participant Client
+    participant GitServer
+    participant Database
+    participant HostMachine
+
+    Client->>GitServer: git-send-pack repository_name (git push)
+    GitServer->>Database: Is the ssh key in the database?
+    Database-->>GitServer: result
+    alt Authenticated
+        GitServer->>Database: Is the ssh key linked to the repository ?
+        Database-->>GitServer: result
+        alt Authorized
+            GitServer->>HostMachine: Execute git-receive-pack
+            HostMachine-->>GitServer: Push Successful (from host)
+            GitServer-->>Client: Push Successful (continue)
+        else Not Authorized
+            GitServer-->>Client: Connection Closed (Not Authorized)
+        end
+    else Not Authenticated
+        GitServer-->>Client: Connection Closed (Not Authenticated)
+    end
+```
+
+##### Building images
+
+Once the code is finally pushed to the repository, it needs to be built in order to be deployed. Since the strategy we chose is [Docker](https://www.docker.com/), the code needs to be built into a docker image.
```

Comment:
```suggestion
Once the code is finally pushed to the repository, it needs to be built in order to be deployed. Since the strategy we have chosen is [Docker](https://www.docker.com/), the code needs to be built into a Docker image.
```